src/swarm-engine/Swarm/Game/Step/Flood.hs

```diff
@@ -0,0 +1,65 @@
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Implementation of the 'Swarm.Language.Syntax.Volume' command for robots.
+module Swarm.Game.Step.Flood (
+  floodFill,
+) where
+
+import Control.Effect.Lens
+import Control.Monad (filterM)
+import Data.HashSet (HashSet)
+import Data.HashSet qualified as HashSet
+import Swarm.Game.Location
+import Swarm.Game.Step.RobotStepState
+import Swarm.Game.Step.Util (checkMoveFailureUnprivileged)
+import Swarm.Game.Step.Util.Inspect (getNeighborLocs)
+import Swarm.Game.Universe
+
+data FloodParms = FloodParms
+  { theSubworld :: SubworldName
+  , maxVisits :: Int
+  }
+
+floodRecursive ::
```

Comment:
I note that because you *prepend* the `visitableNeighbors` to the `pending` list, you are essentially treating it like a stack, so this is doing a DFS rather than a BFS.  When the robot is enclosed, this doesn't make much difference---the same cells end up being discovered either way.  I guess the difference is that in cases where the robot is *not* enclosed, the DFS visits cells in basically a straight line until it has visited the max number, rather than visiting all the cells in a square region around the robot.  I don't know how much this matters.

Note I'm not saying you should append `visitableNeighbors` to the end of `pending`, that would indeed be very inefficient.  If we really want a BFS we would have to switch to queue data structure for `pending` --- `Data.Sequence` works great as a queue in practice (even though in theory it incurs an extra log factor).